src/core/NgenSimulation.cpp

```diff
@@ -0,0 +1,263 @@
+#include <NgenSimulation.hpp>
+#include <NGenConfig.h>
+#include <Logger.hpp>
+
+#if NGEN_WITH_ROUTING
+#include "bmi/Bmi_Py_Adapter.hpp"
+#endif // NGEN_WITH_ROUTING
+
+#include "parallel_utils.h"
+
+NgenSimulation::NgenSimulation(
+    std::shared_ptr<realization::Formulation_Manager> formulation_manager,
+    std::vector<std::shared_ptr<ngen::Layer>> layers,
+    std::unordered_map<std::string, int> catchment_indexes,
+    std::unordered_map<std::string, int> nexus_indexes,
+    int mpi_rank,
+    int mpi_num_procs
+                               )
+    : catchment_formulation_manager_(formulation_manager)
+    , simulation_step_(0)
+    , sim_time_(std::make_shared<Simulation_Time>(*formulation_manager->Simulation_Time_Object))
+    , layers_(std::move(layers))
+    , catchment_indexes_(std::move(catchment_indexes))
+    , nexus_indexes_(std::move(nexus_indexes))
+    , mpi_rank_(mpi_rank)
+    , mpi_num_procs_(mpi_num_procs)
+{
+    catchment_outflows_.reserve(catchment_indexes_.size() * get_num_output_times());
+    nexus_downstream_flows_.reserve(nexus_indexes_.size() * get_num_output_times());
+}
+
+NgenSimulation::~NgenSimulation() = default;
+
+void NgenSimulation::run_catchments()
+{
+    // Now loop some time, iterate catchments, do stuff for total number of output times
+    auto num_times = get_num_output_times();
+
+    for (; simulation_step_ < num_times; simulation_step_++) {
+        // Make room for this output step's results
+        catchment_outflows_.resize(catchment_outflows_.size() + catchment_indexes_.size());
+        nexus_downstream_flows_.resize(nexus_downstream_flows_.size() + nexus_indexes_.size());
```

Comment:
Is the shift to resizing each step to reduce the size of the saved state? A couple considerations given this change:
1) This logic also seems better to include in `advance_one_output_step` instead of here, .e.g, something could call the advance function but the size of the output vectors is not large enough to contain the results.
2) We may want to explicitly set the last `nexus_indxexes_.size()` elements to 0. Since the layer will add the nexus value instead of set it, backtracking in time and decreasing size could yield double counts of that time step.
3) `run_routing` should check the size of `nexus_downstream_flows_`. It was built assuming the size of the data would be correct, but I'm not sure it's safe to do that anymore.

Reply:
Re 2: Good catch, I'll add the explicit initialization argument